File: material.cc
#include "cross_sections.cc"
#include <cstdlib>
#include <vector>

#ifndef MAT
#define MAT

struct Atom {
  Atom(const int a0, const int z0, const std::string ne_r,
       const std::string el_ruth_cs, const std::string ne_ea,
       const double cutoff)
      : a(a0), z(z0), el_ruth_rate(el_ruth_cs, cutoff), ne_rate(ne_r),
        el_ruth_angle_cdf(el_ruth_cs, cutoff), ne_energy_angle(ne_ea) {}

  // Constructor for zero non-elastic rate for hydrogen
  Atom(const int a0, const int z0, const std::string el_ruth_cs,
       const double cutoff, const double back_cutoff)
      : a(a0), z(z0), el_ruth_rate(el_ruth_cs, cutoff, back_cutoff), ne_rate(),
        el_ruth_angle_cdf(el_ruth_cs, cutoff, back_cutoff), ne_energy_angle() {}

  Atom(const Atom &other)
      : a(other.a), z(other.z), el_ruth_rate(other.el_ruth_rate),
        ne_rate(other.ne_rate), el_ruth_angle_cdf(other.el_ruth_angle_cdf),
        ne_energy_angle(other.ne_energy_angle) {}

  double s() const {
    double a_c = a + 1;
    double n_c = a - z;
    double z_c = z + 1;
    double a_a = a;
    double n_a = a - z;
    double z_a = z;
    double ret = 15.68 * (a_c - a_a) -
                 28.07 * (pow(n_c - z_c, 2) / a_c - pow(n_a - z_a, 2) / a_a) -
                 18.56 * (pow(a_c, 2 / 3) - pow(a_a, 2 / 3)) +
                 33.22 * (pow(n_c - z_c, 2) / pow(a_c, 4.0 / 3) -
                          pow(n_a - z_a, 2) / pow(a_a, 4.0 / 3)) -
                 0.717 * (z_c * z_c / pow(a_c, 1.0 / 3) -
                          z_a * z_a / pow(a_a, 1.0 / 3)) +
                 1.211 * (z_c * z_c / a_c - z_a * z_a / a_a);
    return ret;
  }

  void sample_nonelastic_collision(double &e, double &alpha,
                                   gsl_rng *gen) const {
    double out_rvalue, out_energy_cm;
    ne_energy_angle.sample(e, out_rvalue, out_energy_cm, gen);
    double eps_a = a * e / (a + 1);
    double eps_b = (a + 1) * out_energy_cm / a;
    double e_a = eps_a + s();
    double e_b = eps_b + s();
    double x1 = fmin(e_a, 130) * e_b / e_a;
    double x3 = fmin(e_a, 41) * e_b / e_a;
    double aval = 0.04 * x1 + 1.8 * 1e-6 * pow(x1, 3) + 6.7 * 1e-7 * pow(x3, 4);
    double cdfc2 = out_rvalue * cosh(aval) - sinh(aval);
    double cdfc1 = 2 * sinh(aval);
    double u2 = gsl_rng_uniform(gen);
    double z1 = cdfc1 * u2 + cdfc2;
    double z2 =
        (z1 + sqrt(pow(z, 2) - pow(out_rvalue, 2) + 1)) / (out_rvalue + 1);
    double out_angle_cm = log(z2) / aval;
    double out_energy_lab =
        out_energy_cm + e / pow(a + 1, 2) +
        2 * sqrt(out_energy_cm * e) * out_angle_cm / (a + 1);
    double out_angle_lab = sqrt(out_energy_cm / out_energy_lab) * out_angle_cm +
                           sqrt(e / out_energy_lab) / (a + 1);
    e = out_energy_lab;
    alpha = out_angle_lab;
  }

  const int a, z;
  CS_1d el_ruth_rate, ne_rate;
  CS_2d el_ruth_angle_cdf;
  CS_3d ne_energy_angle;
};

struct Material {

  Material(std::vector<Atom> &atoms, const std::vector<int> &id,
           const std::vector<double> &x0, const double d0, const double I0)
      : density(d0), I(I0 / 1e6), x(x0), at() {
    for (unsigned int i = 0; i < id.size(); i++) {
      at.push_back(atoms[id[i]]);
    }
  }

  Material() : density(), I(), x(), at() {}

  void read_material(const std::string filename, std::vector<Atom> &atoms) {
    std::ifstream file;
    file.open(filename + ".txt");
    std::string line, token;
    getline(file, line);
    density = atof(line.c_str());
    getline(file, line);
    I = atof(line.c_str()) / 1e6;
    while (getline(file, line)) {
      std::stringstream iss;
      iss << line;
      getline(iss, token, ' ');
      at.push_back(atoms[atoi(token.c_str())]);
      getline(iss, token, ' ');
      x.push_back(atof(token.c_str()));
    }
    file.close();
    return;
  }

  double bethe_bloch(const double e) const {
    double mecsq = 0.511;   // mass of electron * speed of light squared, MeV
    double mpcsq = 938.346; // mass of proton * speed of light squared, MeV
    double betasq = (2 * mpcsq + e) * e / pow(mpcsq + e, 2);
    double num = 0;
    double denom = 0;
    for (unsigned int i = 0; i < at.size(); i++) {
      num += x[i] * 0.3072 * at[i].z * density *
             (log(2 * mecsq * betasq / (I * (1 - betasq))) - betasq) /
             betasq; // MeV / cm
      denom += at[i].a * x[i];
    }
    return num / denom;
  }

  double multiple_scattering_sd(const double e, const double dt) const {
    double mpcsq = 938.346; // mass of proton * speed of light squared, MeV
    double pv = (2 * mpcsq + e) * e / (mpcsq + e);
    double betasq = (2 * mpcsq + e) * e / pow(mpcsq + e, 2);
    double c = 29979245800; // speed of light
    double vel = sqrt(betasq) * c;
    double p = pv / sqrt(betasq); // momentum in MeV / c.
    // effective chi_c_sq is just the sum of individual elements
    double chi_c_sq = 0;
    std::vector<double> chi_a_sq_vec(at.size());
    for (unsigned int i = 0; i < at.size(); i++) {
      chi_c_sq += x[i] * at[i].z * (at[i].z + 1.0) / at[i].a;
      chi_a_sq_vec[i] = 2.007e-5 * pow(at[i].z, 2 / 3) *
                        (1 + 3.34 * pow(at[i].z / (137 * vel), 2)) / (p * p);
    }
    chi_c_sq *= 0.157 * dt * density / (pv * pv);
    // effective chi_a_sq is a weighted average on the log-scale
    double chi_a_sq = 0;
    double denom = 0;
    for (unsigned int i = 0; i < at.size(); i++) {
      chi_a_sq +=
          x[i] * at[i].z * (at[i].z + 1) * log(chi_a_sq_vec[i]) / at[i].a;
      denom += x[i] * at[i].z * (at[i].z + 1.0) / at[i].a;
    }
    chi_a_sq = exp(chi_a_sq / denom);
    double omega = chi_c_sq / chi_a_sq;
    double F = 0.98;
    double v = omega / (2 * (1 - F));
    double ret = sqrt(chi_c_sq * ((1 + v) * log(1 + v) / v - 1)) / (1 + F * F);
    ret *= dt / sqrt(ret * ret + dt * dt);
    return ret;
  }

  double energy_straggling_sd(const double e) const {
    double alpha = 1 / 137.0;
    double log_hbar = -21 * log(10) + log(4.136) - log(2 * M_PI); // MeV * s
    double log_c = log(29979245800);                              // cm / s
    double log_avogadro = log(6) + 23 * log(10);
    double mpcsq = 938.346; // mass of proton * speed of light squared, MeV
    double betasq = (2 * mpcsq + e) * e / pow(mpcsq + e, 2);
    double a = 0;
    double z = 0;
    for (unsigned int i = 0; i < at.size(); i++) {
      a += x[i] * at[i].a; // average molar mass
      z += x[i] * at[i].z; // electrons per average molecule
    }
    double log_molecule_density =
        log(density) + log_avogadro - log(a); // molecules / cm^3
    double ret =
        4 * M_PI * z * (1 - betasq / 2) / sqrt(1 - betasq) *
        exp(2 * (log(alpha) + log_hbar + log_c) + log_molecule_density);
    return sqrt(ret);
  }

  double nonelastic_rate(const double e) const {
    double log_avogadro = log(6) + 23 * log(10);
    double log_barns_to_cmsq = -24 * log(10);
    double a = 0;
    double ret = 0;
    for (unsigned int i = 0; i < at.size(); i++) {
      a += x[i] * at[i].a; // average molar mass
      ret += at[i].a * x[i] * at[i].ne_rate.evaluate(e);
    }
    double log_molecule_density =
        log(density) + log_avogadro - log(a); // molecules / cm^3
    ret *= exp(log_barns_to_cmsq + log_molecule_density) / a;
    return ret; // rate per cm
  }

  double rutherford_and_elastic_rate(const double e) const {
    double log_avogadro = log(6) + 23 * log(10);
    double log_barns_to_cmsq = -24 * log(10);
    double a = 0;
    double ret = 0;
    for (unsigned int i = 0; i < at.size(); i++) {
      a += x[i] * at[i].a; // average molar mass
      ret += at[i].a * x[i] * at[i].el_ruth_rate.evaluate(e);
    }
    double log_molecule_density =
        log(density) + log_avogadro - log(a); // molecules / cm^3
    ret *= exp(log_barns_to_cmsq + log_molecule_density) / a;
    return ret; // rate per cm
  }

  void compute_new_angle(std::vector<double> &ang, const double alpha,
                         const double beta) {
    double omega_new1 =
        sin(ang[0]) * cos(ang[1]) * cos(alpha) +
        (cos(ang[0]) * cos(ang[1]) * sin(beta) - sin(ang[1]) * cos(beta)) *
            sin(alpha);
    double omega_new2 =
        sin(ang[0]) * sin(ang[1]) * cos(alpha) +
        (cos(ang[0]) * sin(ang[1]) * sin(beta) + cos(ang[1]) * cos(beta)) *
            sin(alpha);
    double omega_new3 =
        cos(ang[0]) * cos(alpha) - sin(ang[0]) * sin(beta) * sin(alpha);
    double magnitude =
        std::sqrt(omega_new1 * omega_new1 + omega_new2 * omega_new2 +
                  omega_new3 * omega_new3);
    omega_new1 /= magnitude;
    omega_new2 /= magnitude;
    omega_new3 /= magnitude;
    ang[0] = acos(omega_new3);
    ang[1] = atan2(omega_new2, omega_new1);
    return;
  }

  void nonelastic_scatter(std::vector<double> &ang, double &e, gsl_rng *gen) {
    double beta = 2 * M_PI * gsl_rng_uniform(gen);
    double rate = 0;
    for (unsigned int i = 0; i < at.size(); i++) {
      rate += at[i].a * x[i] * at[i].ne_rate.evaluate(e);
    }
    double u = gsl_rng_uniform(gen);
    double ind = 0;
    double tmp = at[ind].a * x[ind] * at[ind].ne_rate.evaluate(e) / rate;
    while (tmp < u) {
      ind++;
      tmp += at[ind].a * x[ind] * at[ind].ne_rate.evaluate(e) / rate;
    }
    double alpha;
    // ENDF non-elastic scattering, both energy + angle from CM to LAB
    at[ind].sample_nonelastic_collision(e, alpha, gen);
    alpha = acos(alpha);
    compute_new_angle(ang, alpha, beta);
    return;
  }

  void rutherford_elastic_scatter(std::vector<double> &ang, double &e,
                                  gsl_rng *gen) {
    double beta = 2 * M_PI * gsl_rng_uniform(gen);
    double rate = 0;
    for (unsigned int i = 0; i < at.size(); i++) {
      rate += at[i].a * x[i] * at[i].el_ruth_rate.evaluate(e);
    }
    double u = gsl_rng_uniform(gen);
    double ind = 0;
    double tmp = at[ind].a * x[ind] * at[ind].el_ruth_rate.evaluate(e) / rate;
    while (tmp < u) {
      ind++;
      tmp += at[ind].a * x[ind] * at[ind].el_ruth_rate.evaluate(e) / rate;
    }
    double alpha;
    alpha = at[ind].el_ruth_angle_cdf.sample(e, gen);
    compute_new_angle(ang, alpha, beta);
    return;
  }

  double density; // density, g / cm^3
  double I;       // mean excitation energy, MeV
  std::vector<double> x;
  std::vector<Atom> at;
};

#endif

File: Python/material.py
import math
import numpy as np
import random
from cross_sections import cross_section_1d
from cross_sections import cross_section_2d
from cross_sections import cross_section_3d
import sys


class atom:

    def __init__(self, a0, z0, el_r, ne_r, ne_y, el_a, ne_a, ne_e):
        self.a = a0
        self.z = z0
        self.el_rate = cross_section_1d(el_r)
        self.ne_rate = cross_section_1d(ne_r)
        self.ne_yield = cross_section_1d(ne_y)
        self.el_angle_cdf = cross_section_2d(el_a)
        self.ne_angle_cdf = cross_section_2d(ne_a)
        self.ne_energy_cdf = cross_section_3d(ne_e)

    def cm_to_lab_frame(self, ang, e0, e_delta):
        mp = 938.346  # mass of proton * c^2, MeV
        mn = mp * self.a  # mass of colliding nucleus * c^2, MeV
        u = math.sqrt(e0 * (e0 + mp)) / (e0 + mp + mn)
        g = 1 / math.sqrt(1 - u * u)
        p = math.sqrt((e0 - e_delta) * (e0 - e_delta + 2 * mp))
        e = e0 - e_delta + mp
        v_ratio = u * (e - u * p) / (p - u * e)
        return math.atan(math.sin(ang) / (g * (math.cos(ang) + v_ratio)))


class material:

    def __init__(self, path, atoms):
        data = np.loadtxt(path, max_rows=2)
        self.density = data[0]
        self.I = data[1] / 1e6
        data = np.loadtxt(path, skiprows=2)
        self.x = np.zeros(data.shape[0])
        self.at = []
        for i in range(data.shape[0]):
            self.at.append(atoms[int(data[i, 0])])
            self.x[i] = data[i, 1]

    def bethe_bloch(self, e0):
        mecsq = 0.511  # mass of electron * speed of light squared, MeV
        mpcsq = 938.346  # mass of proton * speed of light squared, MeV
        betasq = (2 * mpcsq + e0) * e0 / (mpcsq + e0) ** 2
        num = (
            0.3072
            * self.density
            * (math.log(2 * mecsq * betasq / self.I * (1 - betasq)) - betasq)
            * sum([self.x[i] * self.at[i].z for i in range(len(self.at))])
            / betasq
        )
        denom = sum([self.at[i].a * self.x[i] for i in range(len(self.at))])
        return num / denom

    def multiple_scattering_sd(self, e0, dt):
        mpcsq = 938.346  # mass of proton * speed of light squared, MeV
        pv = (2 * mpcsq + e0) * e0 / (mpcsq + e0)
        betasq = (2 * mpcsq + e0) * e0 / (mpcsq + e0) ** 2
        c = 29979245800  # speed of light, cm / s
        vel = math.sqrt(betasq) * c
        p = pv / math.sqrt(betasq)  # momentum in MeV / c.
        # effective chi_c_sq is just the sum of individual elements
        chi_c_sq = (
            sum(
                [
                    self.x[i] * self.at[i].z * (self.at[i].z + 1) / self.at[i].a
                    for i in range(len(self.at))
                ]
            )
            * 0.157
            * dt
            * self.density
            / (pv * pv)
        )
        # effective chi_a_sq is a weighted average on the log-scale
        chi_a_sq_vec = [
            2.007e-5
            * self.at[i].z ** (2 / 3)
            * (1 + 3.34 * (self.at[i].z / (137 * vel)) ** 2)
            / (p * p)
            for i in range(len(self.at))
        ]
        chi_a_sq = sum(
            [
                self.x[i]
                * self.at[i].z
                * (self.at[i].z + 1)
                * math.log(chi_a_sq_vec[i])
                / self.at[i].a
                for i in range(len(self.at))
            ]
        )
        denom = sum(
            [
                self.x[i] * self.at[i].z * (self.at[i].z + 1.0) / self.at[i].a
                for i in range(len(self.at))
            ]
        )
        chi_a_sq = math.exp(chi_a_sq / denom)
        omega = chi_c_sq / chi_a_sq
        F = 0.98
        v = omega / (2 * (1 - F))
        ret = math.sqrt(chi_c_sq * ((1 + v) * math.log(1 + v) / v - 1)) / (1 + F * F)
        ret = ret * dt / math.sqrt(ret * ret + dt * dt)
        return ret

    def energy_straggling_sd(self, e0):
        alpha = 1 / 137.0
        log_hbar = (
            -21 * math.log(10) + math.log(4.136) - math.log(2 * math.pi)
        )  # MeV * s
        log_c = math.log(29979245800)  # cm / s
        log_avogadro = math.log(6) + 23 * math.log(10)
        mpcsq = 938.346  # mass of proton * speed of light squared, MeV
        betasq = (2 * mpcsq + e0) * e0 / (mpcsq + e0) ** 2
        a = sum(
            [self.x[i] * self.at[i].a for i in range(len(self.at))]
        )  # average molar mass
        z = sum(
            [self.x[i] * self.at[i].z for i in range(len(self.at))]
        )  # average electrons per molecule
        log_molecule_density = (
            math.log(self.density) + log_avogadro - math.log(a)
        )  # molecules / cm^3
        ret = (
            4
            * math.pi
            * z
            * (1 - betasq / 2)
            / math.sqrt(1 - betasq)
            * math.exp(2 * (math.log(alpha) + log_hbar + log_c) + log_molecule_density)
        )
        return math.sqrt(ret)

    def nonelastic_rate(self, e0):
        log_avogadro = math.log(6) + 23 * math.log(10)
        log_barns_to_cmsq = -24 * math.log(10)
        a = sum(
            [self.x[i] * self.at[i].a for i in range(len(self.at))]
        )  # average molar mass
        ret = sum(
            [
                self.at[i].a * self.x[i] * self.at[i].ne_rate.evaluate(e0)
                for i in range(len(self.at))
            ]
        )
        log_molecule_density = (
            math.log(self.density) + log_avogadro - math.log(a)
        )  # molecules / cm^3
        ret *= math.exp(log_barns_to_cmsq + log_molecule_density) / a
        return ret  # rate per cm

    def elastic_rate(self, e0):
        log_avogadro = math.log(6) + 23 * math.log(10)
        log_barns_to_cmsq = -24 * math.log(10)
        a = sum(
            [self.x[i] * self.at[i].a for i in range(len(self.at))]
        )  # average molar mass
        ret = sum(
            [
                self.at[i].a * self.x[i] * self.at[i].el_rate.evaluate(e0)
                for i in range(len(self.at))
            ]
        )
        log_molecule_density = (
            math.log(self.density) + log_avogadro - math.log(a)
        )  # molecules / cm^3
        ret *= math.exp(log_barns_to_cmsq + log_molecule_density) / a
        return ret  # rate per cm

    def rutherford_rate(self, e0, lb):
        log_ahbarc = math.log(197.3 / 137) - 13 * math.log(10)
        log_avogadro = math.log(6) + 23 * math.log(10)
        mpcsq = 938.346  # mass of proton * speed of light squared, MeV
        pv = (2 * mpcsq + e0) * e0 / (mpcsq + e0)
        a = sum(
            [self.x[i] * self.at[i].a for i in range(len(self.at))]
        )  # average molar mass
        az = sum(
            [
                self.x[i] * self.at[i].a * self.at[i].z * self.at[i].z
                for i in range(len(self.at))
            ]
        )
        log_molecule_density = (
            math.log(self.density) + log_avogadro - math.log(a)
        )  # molecules / cm^3
        sig = (
            math.exp(
                2
                * (
                    log_ahbarc
                    + math.log(math.cos(lb / 2))
                    - math.log(pv)
                    - math.log(math.sin(lb / 2))
                )
                + log_molecule_density
            )
            * math.pi
        )
        ret = sig * az / a
        return ret

    def nonelastic_scatter(self, ang0, e0):
        beta = 2 * math.pi * np.random.uniform()
        rate = sum(
            [
                self.at[i].a * self.x[i] * self.at[i].ne_rate.evaluate(e0)
                for i in range(len(self.at))
            ]
        )
        u = np.random.uniform()
        ind = 0
        tmp = self.at[ind].a * self.x[ind] * self.at[ind].ne_rate.evaluate(e0) / rate
        while tmp < u:
            ind = ind + 1
            tmp = (
                tmp
                + self.at[ind].a
                * self.x[ind]
                * self.at[ind].ne_rate.evaluate(e0)
                / rate
            )
        if np.random.uniform() > self.at[ind].ne_yield.evaluate(e0):
            # proton absorbed & track ends
            ang = 0
            energy = 0
        else:
            alpha = self.at[ind].ne_angle_cdf.sample(e0)
            energy = self.at[ind].ne_energy_cdf.sample(e0, alpha)
            alpha = self.at[ind].cm_to_lab_frame(alpha, e0, e0 - energy)
            ang = ang0
            if 0 <= beta and beta < math.pi / 2:
                ang[0] -= math.atan(math.sin(beta) * math.tan(alpha))
                ang[1] -= math.atan(math.cos(beta) * math.tan(alpha))
            elif math.pi / 2 <= beta and beta < math.pi:
                ang[0] -= math.atan(math.sin(math.pi - beta) * math.tan(alpha))
                ang[1] += math.atan(math.cos(math.pi - beta) * math.tan(alpha))
            elif math.pi <= beta and beta < 3 * math.pi / 2:
                ang[0] += math.atan(math.sin(3 * math.pi / 2 - beta) * math.tan(alpha))
                ang[1] += math.atan(math.cos(3 * math.pi / 2 - beta) * math.tan(alpha))
            else:
                ang[0] += math.atan(math.sin(2 * math.pi - beta) * math.tan(alpha))
                ang[1] -= math.atan(math.cos(2 * math.pi - beta) * math.tan(alpha))
        return ang, energy

    def elastic_scatter(self, ang0, e0):
        beta = 2 * math.pi * np.random.uniform()
        rate = sum(
            [
                self.at[i].a * self.x[i] * self.at[i].el_rate.evaluate(e0)
                for i in range(len(self.at))
            ]
        )
        u = np.random.uniform()
        ind = 0
        tmp = self.at[ind].a * self.x[ind] * self.at[ind].el_rate.evaluate(e0) / rate
        while tmp < u:
            ind = ind + 1
            tmp = (
                tmp
                + self.at[ind].a
                * self.x[ind]
                * self.at[ind].el_rate.evaluate(e0)
                / rate
            )
        alpha = self.at[ind].el_angle_cdf.sample(e0)
        alpha = self.at[ind].cm_to_lab_frame(alpha, e0, 0)
        ang = ang0
        if 0 <= beta and beta < math.pi / 2:
            ang[0] -= math.atan(math.sin(beta) * math.tan(alpha))
            ang[1] -= math.atan(math.cos(beta) * math.tan(alpha))
        elif math.pi / 2 <= beta and beta < math.pi:
            ang[0] -= math.atan(math.sin(math.pi - beta) * math.tan(alpha))
            ang[1] += math.atan(math.cos(math.pi - beta) * math.tan(alpha))
        elif math.pi <= beta and beta < 3 * math.pi / 2:
            ang[0] += math.atan(math.sin(3 * math.pi / 2 - beta) * math.tan(alpha))
            ang[1] += math.atan(math.cos(3 * math.pi / 2 - beta) * math.tan(alpha))
        else:
            ang[0] += math.atan(math.sin(2 * math.pi - beta) * math.tan(alpha))
            ang[1] -= math.atan(math.cos(2 * math.pi - beta) * math.tan(alpha))
        return ang

    def rutherford_scatter(self, ang0, lb):
        beta = 2 * math.pi * np.random.uniform()
        u = np.random.uniform()
        alpha = math.acos(
            (math.cos(lb / 2) ** 2 * (1 - u) - math.sin(lb / 2) ** 2)
            / (1 - u * math.cos(lb / 2) ** 2))
        )
        ang = ang0
        if 0 <= beta and beta < math.pi / 2:
            ang[0] -= math.atan(math.sin(beta) * math.tan(alpha))
            ang[1] -= math.atan(math.cos(beta) * math.tan(alpha))
        elif math.pi / 2 <= beta and beta < math.pi:
            ang[0] -= math.atan(math.sin(math.pi - beta) * math.tan(alpha))
            ang[1] += math.atan(math.cos(math.pi - beta) * math.tan(alpha))
        elif math.pi <= beta and beta < 3 * math.pi / 2:
            ang[0] += math.atan(math.sin(3 * math.pi / 2 - beta) * math.tan(alpha))
            ang[1] += math.atan(math.cos(3 * math.pi / 2 - beta) * math.tan(alpha))
        else:
            ang[0] += math.atan(math.sin(2 * math.pi - beta) * math.tan(alpha))
            ang[1] -= math.atan(math.cos(2 * math.pi - beta) * math.tan(alpha))
        return ang
